src/protobuf/message.cr

```diff
@@ -52,6 +52,16 @@ module Protobuf
         new(::Protobuf::Buffer.new(io))
       end
 
+      # `JSON::Serializable` populates our object without
+      # calling the setters and then invokes `after_initialize`.
+      # We do want our setters to be called (e.g. for `oneof`-handling)
+      # and so we do that here.
+      def after_initialize
+        {% for tag, field in FIELDS %}
+          self.{{field[:name].id}} = @{{field[:name].id}}
+        {% end %}
+      end
```

Comment:
Before this change:

`FooBar.from_json # => { oo_a: 1, oo_b: nil, my_oneof: nil }`

After this change:

`FooBar.from_json # => { oo_a: 1, oo_b: nil, my_oneof: "oo_a" }`

Tested it manually a bunch and it appears to work reliably also with nested objects.